### PR TITLE
Convert to endpoints

### PR DESCRIPTION
**Pull request by johnsca**
Convert to use the Endpoint pattern from juju-solutions/charms.reactive#123

This shouldn't be merged until after that PR.

### CODE REVIEW

**Comment by johnsca** [comment]
Upstream PR has been merged and released as an RC, but we should hold off on this PR until it's an official release.

**Review by kwmonroe — CHANGES_REQUESTED** [review]
I don't think `Endpoint.conversations()` is a thing.

Edit: gah, i can't seem to leave an inline comment on out-of-scope code.  Check around line 44 of requires.py:
```
        for conv in self.conversations():
```

**Review by kwmonroe — APPROVED** [review]
thx, lgtm

**Comment by johnsca** [comment]
@stub Would you mind testing this in a real world charm before I merge this?

**Comment by xannz** [comment]
The provides side seems to break because of PR #12. There is no `relation_name` in the Endpoint class. 
```
2018-06-12 09:18:31 ERROR juju-log Hook error:
Traceback (most recent call last):
  File "/var/lib/juju/agents/unit-gateway-0/.venv/lib/python3.5/site-packages/charms/reactive/__init__.py", line 72, in main
    bus.dispatch(restricted=restricted_mode)
  File "/var/lib/juju/agents/unit-gateway-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 375, in dispatch
    _invoke(other_handlers)
  File "/var/lib/juju/agents/unit-gateway-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 351, in _invoke
    handler.invoke()
  File "/var/lib/juju/agents/unit-gateway-0/.venv/lib/python3.5/site-packages/charms/reactive/bus.py", line 173, in invoke
    self._action(*args)
  File "/var/lib/juju/agents/unit-gateway-0/charm/reactive/nginx-api-gateway.py", line 28, in configure_gateway
    website.configure(port=config().get('port'))
  File "/var/lib/juju/agents/unit-gateway-0/charm/hooks/relations/http/provides.py", line 33, in configure
    hostname = self.get_ingress_address()
  File "/var/lib/juju/agents/unit-gateway-0/charm/hooks/relations/http/provides.py", line 19, in get_ingress_address
    network_info = hookenv.network_get(self.relation_name)
AttributeError: 'HttpProvides' object has no attribute 'relation_name'
```

**Comment by johnsca** [comment]
@xannz Can you test this again?

**Comment by xannz** [comment]
The provides side seems fine but the require still has some issues.
https://github.com/juju-solutions/interface-http/blob/1946df11ac6c106029387324090721cc18db38e2/requires.py#L46-L48
This can not work since `joined_units` keys are unit names and should probably become 
```
private_address = relation.joined_units.received['private-address']
host = relation.joined_units.received['hostname'] or private_address
port = relation.joined_units.received['port']
```

**Comment by johnsca** [comment]
@xannz Thanks.  Fixed.

**Comment by tbaumann** [comment]
I would love to see this merged.
Because the new reactive way is better, and the http layer is always used as an example. This should be a shining example.

:smiley:

File: provides.py
Comment (kwmonroe) [inline provides.py:44]: `set_remote` is not an Endpoint method, yet some charms that implement this interface rely on it, e.g. graylog:

https://git.launchpad.net/graylog-charm/tree/reactive/graylog.py#n508

Is there a back compat way we can handle this either in the Endpoint class or on a per-interface level?

Comment (kwmonroe) [reply provides.py:44]: If we don't handle `set_remote` at the Endpoints class level, I think we'll at least need to allow for charms that want to interact with reverse proxies like haproxy (which are not reactive and rely on `relation-get`), and for that, we'll need to provide more than the current host/ip/port.  Specifically, `services` relation data:

https://bazaar.launchpad.net/~haproxy-team/charm-haproxy/trunk/view/head:/hooks/hooks.py#L599

It doesn't have to be in `configure`. Maybe it could be in -- just spit ballin here -- a new `set_remote` func? ;)

**Review by kwmonroe — APPROVED** [review]

**Pull request merged**
